d2/src/main/java/com/linkedin/d2/xds/XdsClientImpl.java

```diff
@@ -0,0 +1,709 @@
+import io.envoyproxy.envoy.config.listener.v3.Filter;
+import io.envoyproxy.envoy.config.listener.v3.FilterChain;
+import io.envoyproxy.envoy.config.listener.v3.Listener;
+import io.envoyproxy.envoy.config.route.v3.Route;
+import io.envoyproxy.envoy.config.route.v3.RouteConfiguration;
+import io.envoyproxy.envoy.config.route.v3.VirtualHost;
+import io.envoyproxy.envoy.extensions.filters.network.http_connection_manager.v3.HttpConnectionManager;
+import io.envoyproxy.envoy.service.discovery.v3.AggregatedDiscoveryServiceGrpc;
+import io.envoyproxy.envoy.service.discovery.v3.DeltaDiscoveryRequest;
+import io.envoyproxy.envoy.service.discovery.v3.DeltaDiscoveryResponse;
+import io.envoyproxy.envoy.service.discovery.v3.Resource;
+import io.grpc.ManagedChannel;
+import io.grpc.Status;
+import io.grpc.stub.StreamObserver;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.concurrent.ExecutorService;
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+public class XdsClientImpl extends XdsClient
+{
+  private static final String TYPE_URL_HTTP_CONNECTION_MANAGER =
+      "type.googleapis.com/envoy.extensions.filters.network.http_connection_manager.v3.HttpConnectionManager";
+  private static final Logger _log = LoggerFactory.getLogger(XdsClientImpl.class);
+  private static final String ADS_TYPE_URL_LDS = "type.googleapis.com/envoy.config.listener.v3.Listener";
+  private static final String ADS_TYPE_URL_RDS = "type.googleapis.com/envoy.config.route.v3.RouteConfiguration";
+  private static final String ADS_TYPE_URL_CDS = "type.googleapis.com/envoy.config.cluster.v3.Cluster";
+  private static final String ADS_TYPE_URL_EDS = "type.googleapis.com/envoy.config.endpoint.v3.ClusterLoadAssignment";
+
+  private static final String D2_CLUSTER_PROPERTIES_KEY = "d2ClusterProperties";
+  private static final String D2_SERVICE_PROPERTIES_KEY = "d2ServiceProperties";
+
+  private final Node _node;
+  private final ManagedChannel _managedChannel;
+  private final ExecutorService _executorService;
+
+  private final Map<String, ResourceSubscriber> _ldsResourceSubscribers = new HashMap<>();
+  private final Map<String, ResourceSubscriber> _rdsResourceSubscribers = new HashMap<>();
+  private final Map<String, ResourceSubscriber> _cdsResourceSubscribers = new HashMap<>();
+  private final Map<String, ResourceSubscriber> _edsResourceSubscribers = new HashMap<>();
+
+  private AdsStream _adsStream;
+
+  public XdsClientImpl(Node node, ManagedChannel managedChannel, ExecutorService executorService)
+  {
+    _node = node;
+    _managedChannel = managedChannel;
+    _executorService = executorService;
+  }
+
+  private void startRpcStream()
```

Comment:
Should add a TODO here to handle reconnections. An efficient xds client should tell the server its cached resources and versions upon reconnection.

Reply:
There's already a TODO for handling reconnection in AdsStream.handleStreamClosed

Reply:
We also need to handle the cases where a stream cannot be started -- we will need to find the next available Istiod in the future.

Reply:
I think in the future we also need to add a parameter "bool restart", and if set to true, will close the current stream, find the next avail istiod and start a new stream. We don't have to do that in this iteration, yet a TODO should be added, this is useful when the client detects that an Istiod has been unresponsive for a while, and therefore it can choose to connect to another one.

Reply:
Also, in order to handle reconnections more efficiently, xDS client will need to send the initial versions of its cached resources to the xds server. Again, this can be done in the future iteration.